File: src/units/Farmer.scala
package units
import battlecode.common._
import utils.Current.{I, Here}
import utils.Movement.{tryMove, waitMove, wanderUntil}
import utils.Execution.endlessly
import utils.Winning.tryInstantWin
import utils.Farming.{waitPlant, waitWater}
//import utils.Pathfinding.moveAwayFromObstacles
import utils.Constants._

/**
  * Gardener that creates and maintains farms.
  *
  * Invariant: Will never build robots
  * Invariant: When it has created its farm, it will stay inside it forever and just maintain it
  */
class Farmer extends RobotUnit {

  val FARM_RADIUS = 4

  @throws(classOf[GameActionException])
  override def runStep(): Unit = {
    // Find farm location
    System.out.println("Finding farm location...")
    findFarmLocation()
    waitMove(N)

    // Create farm
    System.out.println("Creating farm...")
    circularly {
      waitPlant(_)
    }

    // Maintain farm
    System.out.println("Maintaining farm...")
    endlessly {
      () => {
        tryInstantWin()
        circularly {
          dir => waitWater(Here add(dir, 2), plantIfMissing = true)
        }
      }
    }
  }

  def findFarmLocation(): Unit =
    // Wander until you find a suitable farm location
    while (
      !I.onTheMap(Here, FARM_RADIUS) || I.isCircleOccupiedExceptByThisRobot(Here, FARM_RADIUS)
    )
      System.out.println()
//    tryMove(moveAwayFromObstacles())
    /*wanderUntil(
      condition = () =>
        !I.onTheMap(Here, FARM_RADIUS) || I.isCircleOccupiedExceptByThisRobot(Here, FARM_RADIUS),
      nextDirection = () => {
        moveAwayFromObstacles()
      }
    )*/

  def circularly(action: Direction => Unit): Unit =
    for ((actionDir, moveDir) <- List(
      (N, E), (NE, S), (E, S), (SE, W),
      (S, W), (SW, N), (W, N), (NW, E)
    )) {
      action(actionDir)
      waitMove(moveDir)
    }

  def sense[T](source: List[BodyInfo]): List[Int] = source.map(_ => 1)
}

File: src/units/Farmer.java
package units;

import battlecode.common.Direction;
import battlecode.common.GameActionException;
import battlecode.common.MapLocation;

import java.util.Arrays;
import java.util.List;
import java.util.function.Consumer;

import utils.Movement;
import utils.Pair;

import static utils.Constants.*;
import static utils.Current.Here;
import static utils.Current.I;
import static utils.Execution.*;
import static utils.Movement.*;
import static utils.Farming.*;
import static utils.Winning.tryInstantWin;


/**
 * Gardener that creates and maintains farms.
 *
 * Invariant: Will never build robots
 * Invariant: When it has created its farm, it will stay inside it forever and just maintain it
 */
public class Farmer implements RobotUnit {

    private static final int FARM_RADIUS = 4;

    @Override
    public void runStep() throws GameActionException {
        // Find farm location
        System.out.println("Finding farm...");
        findFarm();
        waitMove(N);

        // Create farm
        System.out.println("Creating farm...");
        circularly(dir -> waitPlant(dir, 0));

        // Maintain farm
        System.out.println("Maintaining farm...");
        endlessly(
            () -> {
                tryInstantWin();
                circularly(
                    dir -> {
                        try {
                            waitWater(Here().add(dir, 2), true);
                        } catch (GameActionException e) {
                            e.printStackTrace();
                        }
                    }
                );
            }
        );

    }

    private void findFarm() {
        wanderUntil(
            Farmer::isLocationSuitableForFarming,
            Movement::randomDirection
        );
    }

    private static Boolean isLocationSuitableForFarming() {
        MapLocation here = Here();
        try {
            return I.onTheMap(here, FARM_RADIUS) && !I.isCircleOccupiedExceptByThisRobot(here, FARM_RADIUS);
        } catch (GameActionException e) {
            e.printStackTrace();
            return false;
        }
    }

    private static void circularly(Consumer<Direction> action) {
        List<Pair<Direction, Direction>> directions = Arrays.asList(
                new Pair<>(N, E),
                new Pair<>(NE, S),
                new Pair<>(E, S),
                new Pair<>(SE, W),
                new Pair<>(S, W),
                new Pair<>(SW, N),
                new Pair<>(W, N),
                new Pair<>(NW, E)
        );
        for (Pair<Direction, Direction> entry : directions) {
            action.accept(entry.fst);
            waitMove(entry.snd);
        }
    }
}
